File: flowersec-swift/Sources/Flowersec/Reconnect.swift
import Foundation

public enum ArtifactSourceKind: String, Equatable, Sendable {
  case once
  case refreshable
}

public struct ArtifactAcquireContext: Equatable, Sendable {
  public var traceID: String?

  public init(traceID: String? = nil) {
    self.traceID = traceID
  }
}

public enum ArtifactSourceError: LocalizedError, Equatable, Sendable {
  case onceConsumed
  case canceled
  case acquisitionFailed(String)

  public var errorDescription: String? {
    switch self {
    case .onceConsumed:
      return "The one-time artifact source has already been consumed."
    case .canceled:
      return "Artifact acquisition was canceled."
    case .acquisitionFailed(let message):
      return "Artifact acquisition failed: \(message)"
    }
  }
}

public struct ArtifactSource: Sendable {
  public let kind: ArtifactSourceKind
  private let acquireHandler: @Sendable (ArtifactAcquireContext) async throws -> ConnectArtifact

  private init(
    kind: ArtifactSourceKind,
    acquire: @escaping @Sendable (ArtifactAcquireContext) async throws -> ConnectArtifact
  ) {
    self.kind = kind
    self.acquireHandler = acquire
  }

  public static func once(_ artifact: ConnectArtifact) -> ArtifactSource {
    let storage = OnceArtifactStorage(artifact)
    return ArtifactSource(kind: .once) { _ in
      try Task.checkCancellation()
      return try await storage.take()
    }
  }

  public static func refreshable(
    _ acquire: @escaping @Sendable (ArtifactAcquireContext) async throws -> ConnectArtifact
  ) -> ArtifactSource {
    ArtifactSource(kind: .refreshable, acquire: acquire)
  }

  public static func controlplane(_ options: ArtifactRequestOptions) -> ArtifactSource {
    .refreshable { context in
      var request = options
      if let traceID = context.traceID { request.traceID = traceID }
      do {
        return try await Controlplane.requestConnectArtifact(request)
      } catch is CancellationError {
        throw ArtifactSourceError.canceled
      } catch {
        throw ArtifactSourceError.acquisitionFailed(error.localizedDescription)
      }
    }
  }

  public static func entryControlplane(
    _ options: ArtifactRequestOptions,
    entryTicket: String
  ) -> ArtifactSource {
    .refreshable { context in
      var request = options
      if let traceID = context.traceID { request.traceID = traceID }
      do {
        return try await Controlplane.requestEntryConnectArtifact(
          request,
          entryTicket: entryTicket
        )
      } catch is CancellationError {
        throw ArtifactSourceError.canceled
      } catch {
        throw ArtifactSourceError.acquisitionFailed(error.localizedDescription)
      }
    }
  }

  public func acquire(_ context: ArtifactAcquireContext = ArtifactAcquireContext()) async throws
    -> ConnectArtifact
  {
    do {
      try Task.checkCancellation()
      return try await acquireHandler(context)
    } catch is CancellationError {
      throw ArtifactSourceError.canceled
    }
  }
}

private actor OnceArtifactStorage {
  private var artifact: ConnectArtifact?

  init(_ artifact: ConnectArtifact) {
    self.artifact = artifact
  }

  func take() throws -> ConnectArtifact {
    guard let artifact else { throw ArtifactSourceError.onceConsumed }
    self.artifact = nil
    return artifact
  }
}

public struct ReconnectSettings: Equatable, Sendable {
  public var enabled: Bool
  public var maxAttempts: Int
  public var initialDelay: Duration
  public var maxDelay: Duration
  public var factor: Double
  public var jitterRatio: Double

  public init(
    enabled: Bool = false,
    maxAttempts: Int = FlowersecSDKDefaults.Reconnect.maxAttempts,
    initialDelay: Duration = .milliseconds(
      FlowersecSDKDefaults.Reconnect.initialDelayMilliseconds
    ),
    maxDelay: Duration = .milliseconds(FlowersecSDKDefaults.Reconnect.maxDelayMilliseconds),
    factor: Double = FlowersecSDKDefaults.Reconnect.factor,
    jitterRatio: Double = FlowersecSDKDefaults.Reconnect.jitterRatio
  ) {
    self.enabled = enabled
    self.maxAttempts = maxAttempts
    self.initialDelay = initialDelay
    self.maxDelay = maxDelay
    self.factor = factor
    self.jitterRatio = jitterRatio
  }

  public func delay(forFailedAttemptIndex attemptIndex: Int) -> Duration {
    let initial = reconnectDurationMilliseconds(initialDelay)
    let maximum = reconnectDurationMilliseconds(maxDelay)
    let exponent = Double(max(0, attemptIndex))
    let base = min(maximum, initial * pow(factor, exponent))
    let jitter = jitterRatio == 0 ? 0 : Double.random(in: -jitterRatio...jitterRatio)
    let milliseconds = max(0, min(Double(Int64.max), base * (1 + jitter)))
    return .milliseconds(Int64(milliseconds.rounded()))
  }

  func normalized() throws -> ReconnectSettings {
    guard maxAttempts >= 1,
      initialDelay >= .zero,
      maxDelay >= .zero,
      factor.isFinite,
      factor >= 1,
      jitterRatio.isFinite,
      (0...1).contains(jitterRatio)
    else { throw ReconnectError.invalidConfiguration }
    var settings = self
    if !enabled { settings.maxAttempts = 1 }
    return settings
  }
}

public enum ReconnectStatus: String, Equatable, Sendable {
  case disconnected
  case connecting
  case connected
  case error
}

public enum ReconnectError: LocalizedError, Equatable, Sendable {
  case refreshableSourceRequired
  case invalidConfiguration
  case canceled
  case artifact(ArtifactSourceError)
  case connection(FlowersecError)
  case terminated(String)
  case exhausted(attempts: Int, last: String)

  public var errorDescription: String? {
    switch self {
    case .refreshableSourceRequired:
      return "Automatic reconnect requires a refreshable artifact source."
    case .invalidConfiguration:
      return "Reconnect configuration is invalid."
    case .canceled:
      return "Reconnect was canceled."
    case .artifact(let error):
      return error.localizedDescription
    case .connection(let error):
      return error.localizedDescription
    case .terminated(let message):
      return message
    case .exhausted(let attempts, let last):
      return "Reconnect attempts exhausted after \(attempts) attempts: \(last)"
    }
  }

  fileprivate var isTerminal: Bool {
    switch self {
    case .refreshableSourceRequired, .invalidConfiguration, .canceled,
      .artifact(.onceConsumed), .artifact(.canceled):
      return true
    case .connection(let error):
      return [
        .invalidInput, .invalidOption, .transportPolicyDenied, .invalidPSK,
        .invalidSuite, .missingConnectInfo, .missingWSURL, .missingChannelID,
        .missingInitExp,
      ].contains(error.code)
    case .artifact(.acquisitionFailed), .terminated, .exhausted:
      return false
    }
  }
}

public struct ReconnectState: Sendable {
  public var status: ReconnectStatus
  public var error: ReconnectError?
  public var client: FlowersecClient?

  public init(
    status: ReconnectStatus = .disconnected,
    error: ReconnectError? = nil,
    client: FlowersecClient? = nil
  ) {
    self.status = status
    self.error = error
    self.client = client
  }
}

typealias ReconnectConnector = @Sendable (ConnectArtifact, ConnectOptions) async throws
  -> FlowersecClient

public struct ReconnectConfig: Sendable {
  public var source: ArtifactSource
  public var options: ConnectOptions
  public var settings: ReconnectSettings
  fileprivate var connector: ReconnectConnector

  public init(
    source: ArtifactSource,
    options: ConnectOptions = ConnectOptions(),
    settings: ReconnectSettings = ReconnectSettings()
  ) {
    self.source = source
    self.options = options
    self.settings = settings
    self.connector = { artifact, options in
      try await Flowersec.connect(artifact, options: options)
    }
  }

  init(
    source: ArtifactSource,
    options: ConnectOptions = ConnectOptions(),
    settings: ReconnectSettings = ReconnectSettings(),
    connector: @escaping ReconnectConnector
  ) {
    self.source = source
    self.options = options
    self.settings = settings
    self.connector = connector
  }
}

public actor ReconnectManager {
  private var currentState = ReconnectState()
  private var generation: UInt64 = 0
  private var supervisor: Task<Void, Never>?
  private var subscribers: [UUID: AsyncStream<ReconnectState>.Continuation] = [:]

  public init() {}

  public func state() -> ReconnectState {
    currentState
  }

  public func subscribe() -> AsyncStream<ReconnectState> {
    let id = UUID()
    let current = currentState
    return AsyncStream { continuation in
      subscribers[id] = continuation
      continuation.yield(current)
      continuation.onTermination = { [weak self] _ in
        Task { await self?.removeSubscriber(id) }
      }
    }
  }

  public func connect(_ config: ReconnectConfig) async throws {
    await stopSupervisor(setDisconnected: false)
    let settings: ReconnectSettings
    do {
      settings = try config.settings.normalized()
      if settings.enabled, config.source.kind != .refreshable {
        throw ReconnectError.refreshableSourceRequired
      }
    } catch let error as ReconnectError {
      setState(.error, error: error, client: nil)
      throw error
    }
    generation &+= 1
    let runGeneration = generation
    setState(.connecting, error: nil, client: nil)
    try await withCheckedThrowingContinuation {
      (continuation: CheckedContinuation<Void, any Error>) in
      supervisor = Task { [weak self] in
        guard let self else {
          continuation.resume(throwing: ReconnectError.canceled)
          return
        }
        await self.run(
          config: config,
          settings: settings,
          generation: runGeneration,
          firstResult: continuation
        )
      }
    }
  }

  public func connectIfNeeded(_ config: ReconnectConfig) async throws {
    switch currentState.status {
    case .connected:
      return
    case .connecting:
      for await state in subscribe() {
        switch state.status {
        case .connected:
          return
        case .error:
          throw state.error ?? ReconnectError.terminated("Connection failed.")
        case .disconnected:
          throw ReconnectError.canceled
        case .connecting:
          continue
        }
      }
      throw ReconnectError.canceled
    case .disconnected, .error:
      try await connect(config)
    }
  }

  public func disconnect() async {
    await stopSupervisor(setDisconnected: true)
  }

  private func stopSupervisor(setDisconnected: Bool) async {
    generation &+= 1
    let task = supervisor
    supervisor = nil
    task?.cancel()
    if let client = currentState.client { await client.close() }
    if let task { await task.value }
    if setDisconnected { setState(.disconnected, error: nil, client: nil) }
  }

  private func run(
    config: ReconnectConfig,
    settings: ReconnectSettings,
    generation runGeneration: UInt64,
    firstResult initialFirstResult: CheckedContinuation<Void, any Error>
  ) async {
    var firstResult: CheckedContinuation<Void, any Error>? = initialFirstResult
    var attemptSequence = 0
    var traceID: String?

    while isActive(runGeneration) {
      var lastError = ReconnectError.terminated("The Flowersec connection terminated unexpectedly.")
      for attempt in 1...settings.maxAttempts {
        guard isActive(runGeneration) else {
          resumeFirst(&firstResult, with: .failure(ReconnectError.canceled))
          return
        }
        attemptSequence += 1
        let started = ContinuousClock.now
        emit(
          code: attempt == 1 ? "reconnect_attempt" : "reconnect_retry_attempt",
          result: .retry,
          options: config.options,
          attemptSequence: attemptSequence,
          traceID: traceID,
          started: started
        )

        let artifact: ConnectArtifact
        do {
          artifact = try await config.source.acquire(ArtifactAcquireContext(traceID: traceID))
          traceID = artifact.metadata.correlation?.traceID ?? traceID
        } catch {
          lastError = reconnectError(error, artifact: true)
          if await finishOrSchedule(
            error: lastError,
            attempt: attempt,
            settings: settings,
            config: config,
            generation: runGeneration,
            attemptSequence: attemptSequence,
            traceID: traceID,
            started: started,
            firstResult: &firstResult
          ) { return }
          continue
        }

        do {
          let client = try await config.connector(
            artifact,
            correlatedOptions(
              config.options,
              attemptSequence: attemptSequence,
              traceID: traceID
            )
          )
          guard isActive(runGeneration) else {
            await client.close()
            resumeFirst(&firstResult, with: .failure(ReconnectError.canceled))
            return
          }
          setState(.connected, error: nil, client: client)
          emit(
            code: "reconnect_connected",
            result: .ok,
            options: config.options,
            attemptSequence: attemptSequence,
            traceID: traceID,
            started: started
          )
          resumeFirst(&firstResult, with: .success(()))
          let termination = await withTaskCancellationHandler {
            await client.terminated()
          } onCancel: {
            Task { await client.close() }
          }
          guard isActive(runGeneration) else { return }
          lastError = .terminated(
            termination?.localizedDescription
              ?? "The Flowersec connection terminated unexpectedly."
          )
          if !settings.enabled {
            setState(.error, error: lastError, client: nil)
            return
          }
          setState(.connecting, error: lastError, client: nil)
          break
        } catch {
          lastError = reconnectError(error, artifact: false)
          if await finishOrSchedule(
            error: lastError,
            attempt: attempt,
            settings: settings,
            config: config,
            generation: runGeneration,
            attemptSequence: attemptSequence,
            traceID: traceID,
            started: started,
            firstResult: &firstResult
          ) { return }
        }
      }
    }
    resumeFirst(&firstResult, with: .failure(ReconnectError.canceled))
  }

  private func finishOrSchedule(
    error: ReconnectError,
    attempt: Int,
    settings: ReconnectSettings,
    config: ReconnectConfig,
    generation runGeneration: UInt64,
    attemptSequence: Int,
    traceID: String?,
    started: ContinuousClock.Instant,
    firstResult: inout CheckedContinuation<Void, any Error>?
  ) async -> Bool {
    let exhausted = error.isTerminal || !settings.enabled || attempt >= settings.maxAttempts
    if exhausted {
      let finalError = error.isTerminal
        ? error
        : ReconnectError.exhausted(attempts: attempt, last: error.localizedDescription)
      setState(.error, error: finalError, client: nil)
      emit(
        code: "reconnect_exhausted",
        result: .fail,
        options: config.options,
        attemptSequence: attemptSequence,
        traceID: traceID,
        started: started
      )
      resumeFirst(&firstResult, with: .failure(finalError))
      return true
    }
    setState(.connecting, error: error, client: nil)
    emit(
      code: "reconnect_scheduled",
      result: .retry,
      options: config.options,
      attemptSequence: attemptSequence,
      traceID: traceID,
      started: started
    )
    do {
      try await Task.sleep(for: settings.delay(forFailedAttemptIndex: attempt - 1))
    } catch {
      resumeFirst(&firstResult, with: .failure(ReconnectError.canceled))
      return true
    }
    return !isActive(runGeneration)
  }

  private func isActive(_ runGeneration: UInt64) -> Bool {
    generation == runGeneration && !Task.isCancelled
  }

  private func setState(
    _ status: ReconnectStatus,
    error: ReconnectError?,
    client: FlowersecClient?
  ) {
    currentState = ReconnectState(status: status, error: error, client: client)
    for continuation in subscribers.values { continuation.yield(currentState) }
  }

  private func removeSubscriber(_ id: UUID) {
    subscribers.removeValue(forKey: id)
  }

  private func emit(
    code: String,
    result: DiagnosticResult,
    options: ConnectOptions,
    attemptSequence: Int,
    traceID: String?,
    started: ContinuousClock.Instant
  ) {
    options.onDiagnosticEvent?(
      DiagnosticEvent(
        path: .auto,
        stage: .reconnect,
        codeDomain: .event,
        code: code,
        result: result,
        elapsedMS: reconnectDurationMilliseconds(started.duration(to: .now)),
        attemptSeq: attemptSequence,
        traceID: traceID
      )
    )
  }
}

private func reconnectError(_ error: any Error, artifact: Bool) -> ReconnectError {
  if let reconnect = error as? ReconnectError { return reconnect }
  if let source = error as? ArtifactSourceError { return .artifact(source) }
  if let flowersec = error as? FlowersecError { return .connection(flowersec) }
  if error is CancellationError { return .canceled }
  if artifact { return .artifact(.acquisitionFailed(error.localizedDescription)) }
  return .terminated(error.localizedDescription)
}

private func correlatedOptions(
  _ options: ConnectOptions,
  attemptSequence: Int,
  traceID: String?
) -> ConnectOptions {
  var resolved = options
  let observer = options.onDiagnosticEvent
  resolved.onDiagnosticEvent = { event in
    var event = event
    event.attemptSeq = attemptSequence
    if event.traceID == nil { event.traceID = traceID }
    observer?(event)
  }
  return resolved
}

private func resumeFirst(
  _ continuation: inout CheckedContinuation<Void, any Error>?,
  with result: Result<Void, any Error>
) {
  guard let current = continuation else { return }
  continuation = nil
  current.resume(with: result)
}

private func reconnectDurationMilliseconds(_ duration: Duration) -> Double {
  let components = duration.components
  return Double(components.seconds) * 1_000
    + Double(components.attoseconds) / 1_000_000_000_000_000
}

File: flowersec-go/reconnect/reconnect.go
// Package reconnect provides supervised reconnect with refreshable connect artifacts.
package reconnect

import (
	"context"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"sync"
	"time"

	"github.com/floegence/flowersec/flowersec-go/client"
	cpclient "github.com/floegence/flowersec/flowersec-go/controlplane/client"
	"github.com/floegence/flowersec/flowersec-go/fserrors"
	"github.com/floegence/flowersec/flowersec-go/internal/defaults"
	"github.com/floegence/flowersec/flowersec-go/observability"
	"github.com/floegence/flowersec/flowersec-go/protocolio"
)

// SourceKind identifies whether a connect artifact can be acquired more than once.
type SourceKind string

const (
	SourceOnce        SourceKind = "once"
	SourceRefreshable SourceKind = "refreshable"
)

// AcquireContext carries correlation state into a fresh artifact request.
type AcquireContext struct {
	TraceID string
}

// ArtifactSource provides one-time or refreshable connect artifacts.
type ArtifactSource interface {
	Kind() SourceKind
	Acquire(context.Context, AcquireContext) (*protocolio.ConnectArtifact, error)
}

type onceSource struct {
	mu       sync.Mutex
	artifact *protocolio.ConnectArtifact
}

// OnceSource creates an artifact source that can be consumed exactly once.
func OnceSource(artifact *protocolio.ConnectArtifact) ArtifactSource {
	return &onceSource{artifact: artifact}
}

func (s *onceSource) Kind() SourceKind { return SourceOnce }

func (s *onceSource) Acquire(ctx context.Context, _ AcquireContext) (*protocolio.ConnectArtifact, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.artifact == nil {
		return nil, errors.New("one-time artifact source has already been consumed")
	}
	artifact := s.artifact
	s.artifact = nil
	return artifact, nil
}

type refreshableSource struct {
	acquire func(context.Context, AcquireContext) (*protocolio.ConnectArtifact, error)
}

// RefreshableSource creates an artifact source backed by an acquisition function.
func RefreshableSource(acquire func(context.Context, AcquireContext) (*protocolio.ConnectArtifact, error)) ArtifactSource {
	return &refreshableSource{acquire: acquire}
}

func (s *refreshableSource) Kind() SourceKind { return SourceRefreshable }

func (s *refreshableSource) Acquire(ctx context.Context, acquireCtx AcquireContext) (*protocolio.ConnectArtifact, error) {
	if s == nil || s.acquire == nil {
		return nil, errors.New("artifact acquisition function is required")
	}
	return s.acquire(ctx, acquireCtx)
}

// ControlplaneSource creates a refreshable source for the standard artifact endpoint.
func ControlplaneSource(config cpclient.ConnectArtifactRequestConfig) ArtifactSource {
	return RefreshableSource(func(ctx context.Context, acquireCtx AcquireContext) (*protocolio.ConnectArtifact, error) {
		config.TraceID = acquireCtx.TraceID
		return cpclient.RequestConnectArtifact(ctx, config)
	})
}

// EntryControlplaneSource creates a refreshable source for the entry-ticket endpoint.
func EntryControlplaneSource(config cpclient.EntryConnectArtifactRequestConfig) ArtifactSource {
	return RefreshableSource(func(ctx context.Context, acquireCtx AcquireContext) (*protocolio.ConnectArtifact, error) {
		config.TraceID = acquireCtx.TraceID
		return cpclient.RequestEntryConnectArtifact(ctx, config)
	})
}

// Settings configure automatic reconnect attempts and exponential backoff.
type Settings struct {
	Enabled      bool
	MaxAttempts  int
	InitialDelay time.Duration
	MaxDelay     time.Duration
	Factor       float64
	JitterRatio  float64
}

// DefaultSettings returns the cross-language canonical reconnect defaults.
func DefaultSettings() Settings {
	return Settings{
		MaxAttempts:  defaults.ReconnectMaxAttempts,
		InitialDelay: defaults.ReconnectInitialDelay,
		MaxDelay:     defaults.ReconnectMaxDelay,
		Factor:       defaults.ReconnectFactor,
		JitterRatio:  defaults.ReconnectJitterRatio,
	}
}

func (s Settings) normalized() (Settings, error) {
	if !s.Enabled {
		s.MaxAttempts = 1
		if s.Factor == 0 {
			s.Factor = 1
		}
	}
	if s.MaxAttempts < 1 || s.InitialDelay < 0 || s.MaxDelay < 0 || s.Factor < 1 || math.IsNaN(s.Factor) || s.JitterRatio < 0 || s.JitterRatio > 1 || math.IsNaN(s.JitterRatio) {
		return Settings{}, errors.New("invalid reconnect settings")
	}
	return s, nil
}

func (s Settings) retryDelay(failedAttemptIndex int) time.Duration {
	base := float64(s.InitialDelay) * math.Pow(s.Factor, float64(failedAttemptIndex))
	if base > float64(s.MaxDelay) {
		base = float64(s.MaxDelay)
	}
	jitter := 0.0
	if s.JitterRatio > 0 {
		jitter = (rand.Float64()*2 - 1) * s.JitterRatio
	}
	delay := base * (1 + jitter)
	if delay < 0 {
		return 0
	}
	return time.Duration(delay)
}

// Status is the stable reconnect lifecycle state.
type Status string

const (
	StatusDisconnected Status = "disconnected"
	StatusConnecting   Status = "connecting"
	StatusConnected    Status = "connected"
	StatusError        Status = "error"
)

// State is a snapshot of reconnect status.
type State struct {
	Status Status
	Error  error
	Client client.Client
}

// Connector performs one high-level Flowersec connection attempt.
type Connector func(context.Context, *protocolio.ConnectArtifact, ...client.ConnectOption) (client.Client, error)

// Config defines a supervised connection.
type Config struct {
	Source         ArtifactSource
	ConnectOptions []client.ConnectOption
	Observer       observability.ClientObserver
	Reconnect      Settings
	Connector      Connector
}

// Manager supervises one active Flowersec client at a time.
type Manager struct {
	mu          sync.Mutex
	state       State
	generation  uint64
	cancel      context.CancelFunc
	done        chan error
	subscribers map[chan State]struct{}
}

// NewManager creates a disconnected reconnect manager.
func NewManager() *Manager {
	return &Manager{
		state:       State{Status: StatusDisconnected},
		subscribers: make(map[chan State]struct{}),
	}
}

// State returns the current reconnect state.
func (m *Manager) State() State {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.state
}

// Subscribe returns a buffered state stream and an unsubscribe function.
func (m *Manager) Subscribe() (<-chan State, func()) {
	ch := make(chan State, 1)
	m.mu.Lock()
	m.subscribers[ch] = struct{}{}
	current := m.state
	m.mu.Unlock()
	ch <- current
	return ch, func() {
		m.mu.Lock()
		if _, ok := m.subscribers[ch]; ok {
			delete(m.subscribers, ch)
			close(ch)
		}
		m.mu.Unlock()
	}
}

// Connect replaces any active connection and waits for the first successful connection or terminal failure.
func (m *Manager) Connect(ctx context.Context, config Config) error {
	if ctx == nil {
		ctx = context.Background()
	}
	if config.Source == nil {
		return errors.New("artifact source is required")
	}
	settings, err := config.Reconnect.normalized()
	if err != nil {
		return err
	}
	if settings.Enabled && config.Source.Kind() != SourceRefreshable {
		return errors.New("automatic reconnect requires a refreshable artifact source")
	}
	if err := m.Disconnect(); err != nil {
		return fmt.Errorf("disconnect existing reconnect session: %w", err)
	}
	runCtx, cancel := context.WithCancel(ctx)
	m.mu.Lock()
	m.generation++
	generation := m.generation
	m.cancel = cancel
	m.done = make(chan error, 1)
	done := m.done
	m.mu.Unlock()
	m.setState(State{Status: StatusConnecting})
	first := make(chan error, 1)
	go func() {
		done <- m.supervise(runCtx, generation, config, settings, first)
		close(done)
	}()
	return <-first
}

// ConnectIfNeeded preserves a healthy or in-progress connection.
func (m *Manager) ConnectIfNeeded(ctx context.Context, config Config) error {
	state := m.State()
	if state.Status == StatusConnected {
		return nil
	}
	if state.Status == StatusConnecting {
		states, unsubscribe := m.Subscribe()
		defer unsubscribe()
		for {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case state, ok := <-states:
				if !ok {
					return errors.New("reconnect state stream closed")
				}
				if state.Status == StatusConnected {
					return nil
				}
				if state.Status == StatusError || state.Status == StatusDisconnected {
					if state.Error != nil {
						return state.Error
					}
					return errors.New("reconnect stopped before connecting")
				}
			}
		}
	}
	return m.Connect(ctx, config)
}

// Disconnect stops retries, joins the supervisor, and closes the active client.
func (m *Manager) Disconnect() error {
	m.mu.Lock()
	cancel := m.cancel
	done := m.done
	active := m.state.Client
	m.cancel = nil
	m.done = nil
	m.generation++
	m.mu.Unlock()
	if cancel != nil {
		cancel()
	}
	var disconnectErr error
	if done != nil {
		disconnectErr = errors.Join(disconnectErr, <-done)
	}
	if active != nil {
		disconnectErr = errors.Join(disconnectErr, active.Close())
	}
	if disconnectErr != nil {
		m.setState(State{Status: StatusError, Error: disconnectErr})
		return disconnectErr
	}
	m.setState(State{Status: StatusDisconnected})
	return nil
}

func (m *Manager) supervise(ctx context.Context, generation uint64, config Config, settings Settings, first chan<- error) error {
	connector := config.Connector
	if connector == nil {
		connector = func(ctx context.Context, artifact *protocolio.ConnectArtifact, options ...client.ConnectOption) (client.Client, error) {
			return client.Connect(ctx, artifact, options...)
		}
	}
	traceID := ""
	firstSent := false
	attemptSeq := 0
	for {
		var lastErr error
		for attempt := 1; attempt <= settings.MaxAttempts; attempt++ {
			if err := ctx.Err(); err != nil {
				if !firstSent {
					first <- err
				}
				return nil
			}
			attemptSeq++
			attemptStart := time.Now()
			observer := observability.WithClientObserverContext(config.Observer, observability.ClientObserverContext{
				Path:         fserrors.PathAuto,
				AttemptSeq:   attemptSeq,
				AttemptStart: attemptStart,
			})
			emit(observer, attemptCode(attempt), observability.DiagnosticResultRetry)
			artifact, err := config.Source.Acquire(ctx, AcquireContext{TraceID: traceID})
			if err == nil && artifact == nil {
				err = errors.New("artifact source returned nil artifact")
			}
			if err == nil && artifact.Correlation != nil && artifact.Correlation.TraceID != nil {
				traceID = *artifact.Correlation.TraceID
				observer = observability.WithClientObserverContext(observer, observability.ClientObserverContext{TraceID: &traceID})
			}
			var connected client.Client
			if err == nil {
				options := append([]client.ConnectOption(nil), config.ConnectOptions...)
				if observer != nil {
					options = append(options, client.WithObserver(observer))
				}
				connected, err = connector(ctx, artifact, options...)
			}
			if err == nil {
				if !m.isCurrent(generation) {
					return connected.Close()
				}
				m.setState(State{Status: StatusConnected, Client: connected})
				emit(observer, "reconnect_connected", observability.DiagnosticResultOK)
				if !firstSent {
					first <- nil
					firstSent = true
				}
				done := clientTermination(connected)
				if done == nil {
					return nil
				}
				select {
				case <-ctx.Done():
					return connected.Close()
				case <-done:
					lastErr = errors.New("Flowersec session terminated")
				}
				_ = connected.Close()
				if !settings.Enabled {
					m.setState(State{Status: StatusError, Error: lastErr})
					return nil
				}
				m.setState(State{Status: StatusConnecting, Error: lastErr})
				emit(observer, "reconnect_scheduled", observability.DiagnosticResultRetry)
				if !waitRetry(ctx, settings.retryDelay(0)) {
					return nil
				}
				break
			}
			lastErr = err
			if isTerminal(err) || attempt == settings.MaxAttempts || !settings.Enabled {
				finalErr := err
				if !isTerminal(err) && settings.Enabled {
					finalErr = fmt.Errorf("reconnect attempts exhausted after %d attempts: %w", attempt, err)
				}
				m.setState(State{Status: StatusError, Error: finalErr})
				emit(observer, "reconnect_exhausted", observability.DiagnosticResultFail)
				if !firstSent {
					first <- finalErr
				}
				return nil
			}
			m.setState(State{Status: StatusConnecting, Error: err})
			emit(observer, "reconnect_scheduled", observability.DiagnosticResultRetry)
			if !waitRetry(ctx, settings.retryDelay(attempt-1)) {
				if !firstSent {
					first <- ctx.Err()
				}
				return nil
			}
		}
		if !settings.Enabled {
			if !firstSent {
				first <- lastErr
			}
			return nil
		}
	}
}

func waitRetry(ctx context.Context, delay time.Duration) bool {
	timer := time.NewTimer(delay)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

func (m *Manager) setState(state State) {
	m.mu.Lock()
	m.state = state
	for subscriber := range m.subscribers {
		select {
		case subscriber <- state:
		default:
			select {
			case <-subscriber:
			default:
			}
			select {
			case subscriber <- state:
			default:
			}
		}
	}
	m.mu.Unlock()
}

func (m *Manager) isCurrent(generation uint64) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.generation == generation
}

func clientTermination(value client.Client) <-chan struct{} {
	internal, ok := value.(client.ClientInternal)
	if !ok || internal.Mux() == nil {
		return nil
	}
	return internal.Mux().CloseChan()
}

func isTerminal(err error) bool {
	if errors.Is(err, context.Canceled) {
		return true
	}
	var flowersecErr *fserrors.Error
	if !errors.As(err, &flowersecErr) {
		return false
	}
	_, ok := terminalConnectCodes[flowersecErr.Code]
	return ok
}

var terminalConnectCodes = map[fserrors.Code]struct{}{
	fserrors.CodeInvalidInput:          {},
	fserrors.CodeInvalidOption:         {},
	fserrors.CodeRoleMismatch:          {},
	fserrors.CodeTransportPolicyDenied: {},
	fserrors.CodeInvalidPSK:            {},
	fserrors.CodeInvalidSuite:          {},
	fserrors.CodeMissingGrant:          {},
	fserrors.CodeMissingConnectInfo:    {},
	fserrors.CodeMissingTunnelURL:      {},
	fserrors.CodeMissingWSURL:          {},
	fserrors.CodeMissingChannelID:      {},
	fserrors.CodeMissingToken:          {},
	fserrors.CodeMissingInitExp:        {},
}

func emit(observer observability.ClientObserver, code string, result observability.DiagnosticResult) {
	if observer == nil {
		return
	}
	observer.OnDiagnosticEvent(observability.DiagnosticEvent{
		Path:       string(fserrors.PathAuto),
		Stage:      observability.DiagnosticStageReconnect,
		CodeDomain: observability.DiagnosticCodeDomainEvent,
		Code:       code,
		Result:     result,
	})
}

func attemptCode(attempt int) string {
	if attempt == 1 {
		return "reconnect_attempt"
	}
	return "reconnect_retry_attempt"
}
